runtime/vm/KeyHashTable.c

```diff
@@ -851,3 +851,58 @@ findClassLocationForClass(J9VMThread *currentThread, J9Class *clazz)
 
 	return targetPtr;
 }
+
+static UDATA
+classLoaderPtrHashFn(void *entry, void *userData)
+{
+	return (UDATA)*(J9ClassLoader **)entry;
+}
+
+static UDATA
+classLoaderPtrHashEqualFn(void *a, void *b, void *userData)
+{
+	return *(J9ClassLoader **)a == *(J9ClassLoader **)b;
+}
+
+static J9HashTable *
+outlivingLoadersTableNew(J9JavaVM *javaVM, U_32 initialSize)
+{
+	return hashTableNew(OMRPORT_FROM_J9PORT(javaVM->portLibrary), J9_GET_CALLSITE(), initialSize, sizeof(J9ClassLoader *), sizeof(J9ClassLoader *), 0, J9MEM_CATEGORY_CLASSES, classLoaderPtrHashFn, classLoaderPtrHashEqualFn, NULL, NULL);
+}
+
+static BOOLEAN
+isLoaderPermanent(J9JavaVM *vm, J9ClassLoader *loader)
+{
+	return loader == vm->systemClassLoader || loader == vm->extensionClassLoader || loader == vm->applicationClassLoader;
+}
+
+void
+addOutlivingLoader(J9VMThread *currentThread, J9ClassLoader *classLoader, J9ClassLoader *outlivingLoader)
+{
+	J9JavaVM *vm = currentThread->javaVM;
+
+	Assert_VM_false(outlivingLoader == classLoader);
+	Assert_VM_mustOwnMonitor(vm->classTableMutex);
+
+	if (!isLoaderPermanent(vm, classLoader) && !isLoaderPermanent(vm, outlivingLoader)) {
+		if (NULL == classLoader->outlivingLoaders) {
+			classLoader->outlivingLoaders = (void *)((UDATA)outlivingLoader | J9CLASSLOADER_OUTLIVING_LOADERS_SINGLE_TAG);
+		} else if (J9_ARE_ANY_BITS_SET((UDATA)classLoader->outlivingLoaders, J9CLASSLOADER_OUTLIVING_LOADERS_SINGLE_TAG)) {
+			J9ClassLoader *existing = (J9ClassLoader *)((UDATA)classLoader->outlivingLoaders & ~(UDATA)J9CLASSLOADER_OUTLIVING_LOADERS_SINGLE_TAG);
+			if (existing != outlivingLoader) {
+				J9HashTable *table = outlivingLoadersTableNew(vm, 2);
+				if (NULL != table) {
+					J9ClassLoader **entry0 = (J9ClassLoader **)hashTableAdd(table, &existing);
+					J9ClassLoader **entry1 = (J9ClassLoader **)hashTableAdd(table, &outlivingLoader);
+					if ((NULL != entry0) && (NULL != entry1)) {
+						classLoader->outlivingLoaders = table;
+					} else {
+						hashTableFree(table);
+					}
+				}
+			}
+		} else {
+			hashTableAdd((J9HashTable *)classLoader->outlivingLoaders, &outlivingLoader);
```

Comment:
Is this an error (adding the address of a local into the hash table)?

Reply:
Or is it going to be copied into the node by the API?

Reply:
I believe it's copied. `&outlivingLoader` is a `J9ClassLoader**`, but the element type is `J9ClassLoader*` (only one level). If somehow this put the local's address into the hash table instead, then the JIT side of my change would surely crash, but I've seen it get the correct results based on the `J9ClassLoader` pointers it got from the hash table

Reply:
Grepping for `hashTableAdd()`, it seems to be common to pass the address of a local variable.